src/lib/optimizer/optimizer.cpp

```diff
@@ -112,7 +114,12 @@ std::shared_ptr<Optimizer> Optimizer::create_default_optimizer() {
   return optimizer;
 }
 
-void Optimizer::add_rule(std::unique_ptr<AbstractRule> rule) { _rules.emplace_back(std::move(rule)); }
+Optimizer::Optimizer(const std::shared_ptr<AbstractCostEstimator>& cost_estimator) : _cost_estimator(cost_estimator) {}
+
+void Optimizer::add_rule(std::unique_ptr<AbstractRule> rule) {
```

Comment:
@mrks alternative would be:

```
template<typename RuleType>
void Optimizer::add_rule() {
  rules.emplace_back(std::make_shared<RuleType>(_cost_estimator);
}
```

Reply:
That will be ugly when rules have parameters. Still, I don't like how an `add_x` method touches `x`. `use_rule` is not too much better, but at least it doesn't have that implicit guarantee that the rule won't be modified.